### PR TITLE
[Test Infra] Full Format Coverage and Sweep for all Tests 

### PR DESCRIPTION
**Pull request by skotaracTT**
### Ticket

https://github.com/orgs/tenstorrent/projects/123/views/1?pane=issue&itemId=112353620&issue=tenstorrent%7Ctt-llk%7C242
### Problem description

Every test now performs a comprehensive sweep across all supported floating-point formats. Before we had certain tests covering certain formats, doing a partial sweep for all supported formats.

Notes:
* Adjustments were necessary for the Bfp8_b format. Once a non-infinite number appears in a row of 16 elements, all numbers in that row get zeroed out (due to sharing the same exponent). This behavior is now correctly reflected in the golden generated tensors as well.
* For fused tests involving Bfp8_b formats, the results produced are correct and accurate; however, precision can drop to about 98% at best. Consequently, we either skip some tests or introduce tolerance thresholds to account for this precision loss. This reduction in precision occurs primarily when copying results from the first L1-to-L1 stage, and is further compounded when truncating values to the Bfp8_b format.

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### CODE REVIEW

File: tests/python_tests/helpers/test_config.py [hunk 1]
```diff
@@ -85,6 +85,16 @@ def generate_build_header(
     unpack_to_dest = str(test_config.get("unpack_to_dest", False)).lower()
     header_content.append(f"#define UNPACKING_TO_DEST {unpack_to_dest}")
 
+    # Tilize / Untilize / Sfpu doesn't work on Tf32 unpacked datums: We unpack 32 bit input to src registers in 16 bit format
```
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/python_tests/helpers/test_config.py:88]: I completely agree with you here! My initial logic was as follows: I have to truncate from 32-bit when unpacking into src registers however in order to preserve precision instead of unpacking to 16-bit for every case, I unpack to Tf32... And in cases where unpacker cannot unpack Tf32 in src regs (tilizing) or SFPU testing, then I unpack to 16-bit datums in src regs... Hence the need for truncate16 flag, and I agree it should be explicitly controlled instead of implicitly but it adds another layer of control for test writers who allow data format conversion to be implicitly handled...
After some thought, I'm not really saving on precision in most cases by unpacking to Tf32, so it might make sense to unpack Float32 to Float16_a/b and have this depend on is_fp32_dest_acc_en as you mentioned 👍

File: tests/helpers/include/data_format_inference.h [hunk 2]
```diff
@@ -42,14 +42,29 @@ constexpr FormatConfig get_data_formats(DataFormat input, DataFormat output, boo
 
     if (input == DataFormat::Float32 && !UNPACKING_TO_DEST)
     {
-        unpack_out = DataFormat::Tf32;
+        if (truncate_16bit)
```
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/helpers/include/data_format_inference.h:45]: Added a comment above to re-iterate our conversation👌

File: tests/python_tests/test_matmul_and_unary_sfpu.py [hunk 3]
```diff
@@ -35,8 +35,8 @@
 from helpers.utils import passed_test
 
 # SUPPORTED FORMATS FOR TEST
-supported_formats = [DataFormat.Float16, DataFormat.Float16_b]
-
+supported_formats = [DataFormat.Float16, DataFormat.Float16_b, DataFormat.Float32]
+supported_formats = [DataFormat.Bfp8_b]
```
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/python_tests/test_matmul_and_unary_sfpu.py:39]: Good catch!😅 so many changed files I got mixed up and forgot to revert changes I made when testing

File: tests/python_tests/test_matmul_pack_untilize.py [hunk 4]
```diff
@@ -26,7 +26,7 @@
 supported_formats = [
     DataFormat.Float16_b,
     DataFormat.Float16,
-    DataFormat.Bfp8_b,
+    DataFormat.Bfp8_b,  # Pack Untilize doesn't work for block float formats (Bfp8_b); we only include as input format in our test
```
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/python_tests/test_matmul_pack_untilize.py:29]: It wasn't a change made in this PR, but I am skipping Bfp8_b. In a previous PR this test was sweeping Bfp8_b and the skip was added then, it doesn't show here since the change happened there.

[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/python_tests/test_matmul_pack_untilize.py:29]: The change on this line was that I added a comment 😁

File: tests/python_tests/test_matmul_unpack_tilize.py [hunk 5]
```diff
@@ -26,7 +26,10 @@
 supported_formats = [
     DataFormat.Float16_b,
     DataFormat.Float16,
-]  # Add DataFormat.Float32 when Data format Inference Model 2.0 supports format conversions for > 1 pipeline run
+    DataFormat.Float32,
+]  #  Add DataFormat.Bfp8_b only as input when Data format Inference Model 2.0 supports format conversions for > 1 pipeline run with different inputs and outputs.
+#  Now tests runs with same input and output format.
+#  We cannot unpack tilize on Bfp8_b format, so it will be included only as input format.
```
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/python_tests/test_matmul_unpack_tilize.py:32]: Yes this is coming in second PR... As the comment suggests we need data format inf for fused tests, right now fused tests run where input and output is same so that intermediate tensors that circulate have the same format in L1 and can be re-used. This will change in PR: https://github.com/tenstorrent/tt-llk/pull/442 and we will be able to sweep across different input and output formats. Thus, for now we cannot only test output Bfp8_b because it required input to also be Bfp8_b 🥲

File: tests/python_tests/helpers/unpack.py [hunk 6]
```diff
@@ -8,9 +8,20 @@
 
 import torch
 
+from helpers.format_arg_mapping import format_dict
+from helpers.format_config import DataFormat
+
 from .format_arg_mapping import DataFormat, format_dict, format_tile_sizes
 
 
+def check_values_in_range(tensor: torch.Tensor, format: DataFormat):
```
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/python_tests/helpers/unpack.py:17]: Good news, I no longer need this check because I setup every sfpu op that returns a non-finite number to model expected non-finite number handling from Tensix. So we do not convert every non-finite number to float("nan") in order for the test to pass. The non-finite cases are resolved according to inputs and formats 👍

File: tests/python_tests/helpers/utils.py [hunk 7]
```diff
@@ -199,4 +213,11 @@ def get_tolerance(output_data_format):
 
     pcc = calculate_pcc(res_tensor, golden_tensor)
 
+    if fused_with_bfp8_b:
```
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/python_tests/helpers/utils.py]: Also resolved, since we agreed to have an accumulated PCC for fused tests where precision loss is accumulated when reusing output of first run L1-L1 as input to the second once (precision loss from intermediate tensors). 
Now the PCC is pcc ^ L1_to_L1_runs

File: tests/helpers/include/data_format_inference.h [hunk 8]
```diff
@@ -42,22 +42,15 @@ constexpr FormatConfig get_data_formats(DataFormat input, DataFormat output, boo
 
     if (input == DataFormat::Float32 && !UNPACKING_TO_DEST)
     {
-        if (truncate_16bit)
+        if (is_exponentB(output) || output == DataFormat::Float32)
```
Comment (lpremovicTT) [inline tests/helpers/include/data_format_inference.h:45]: I think we should still have a path for FP32 to TF32, is that comming in the next PR?

File: tests/python_tests/helpers/golden_generators.py [hunk 9]
```diff
@@ -16,17 +16,18 @@
 golden_registry = {}
 
 
-def check_bfp8_b(operand: torch.Tensor, format: DataFormat) -> None:
+def check_bfp8_b(operand: list) -> list:
```
Comment (lpremovicTT) [inline tests/python_tests/helpers/golden_generators.py:19]: Has pack - unpack magic not helped here and we still need this?

File: tests/python_tests/helpers/utils.py [hunk 10]
```diff
@@ -162,13 +160,8 @@ def passed_test(
     golden_tensor,
     res_tensor,
     output_data_format=DataFormat.Float16_b,
-    fused_with_bfp8_b: bool = False,
+    L1_to_L1_iterations: bool = 1,
```
Comment (lpremovicTT) [inline tests/python_tests/helpers/utils.py:163]: bool?

File: tests/python_tests/helpers/utils.py [hunk 11]
```diff
@@ -199,4 +206,9 @@ def get_tolerance(output_data_format):
 
     pcc = calculate_pcc(res_tensor, golden_tensor)
 
-    return is_within_tolerance and (pcc > 0.99)
+    # Once we iterate L1-L1 more than once the loss in percision is accumulated because the result from the first run is transferred as input to the next run
+    # We don't have a robust accuracy model to determine exact percision loss from each run and accumulate as such per test, so we use a heuristic
+    #   - This reduction in precision occurs primarily when copying results from the first L1-to-L1 stage, and is further compounded when truncating values with less percision (Bfp8_b for example).
+    target_pcc = pow(0.99, L1_to_L1_iterations)
```
Comment (lpremovicTT) [inline tests/python_tests/helpers/utils.py:212]: Should this only be for Bfp8 or all formats?

File: tests/python_tests/test_unpack_untilize.py [hunk 12]
```diff
@@ -22,7 +22,12 @@
 from helpers.utils import passed_test
 
 # SUPPORTED FORMATS FOR TEST
-supported_formats = [DataFormat.Float16, DataFormat.Float16_b]
+supported_formats = [
+    DataFormat.Float16_b,
+    DataFormat.Float16,
+    DataFormat.Float32,
+    DataFormat.Bfp8_b,
```
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/python_tests/test_unpack_untilize.py:29]: Agreed, will keep this in test cases for now and later on we can skip them once we test on more tiles

File: tests/python_tests/helpers/utils.py (hunk 10 above)
Comment (skotaracTT) [reply tests/python_tests/helpers/utils.py:163]: good catch!

File: tests/helpers/include/data_format_inference.h (hunk 8 above)
Comment (skotaracTT) [reply tests/helpers/include/data_format_inference.h]: Done

File: tests/python_tests/helpers/utils.py (hunk 11 above)
Comment (skotaracTT) [reply tests/python_tests/helpers/utils.py]: Done

**Review by lpremovicTT — APPROVED** [review]

**Review by fvranicTT — APPROVED** [review]
